spec/requests/api/v3/work_packages/work_packages_by_project_resource_spec.rb

```diff
@@ -0,0 +1,132 @@
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+
+require 'spec_helper'
+require 'rack/test'
+
+describe API::V3::WorkPackages::WorkPackagesByProjectAPI, type: :request do
+  include Rack::Test::Methods
+  include API::V3::Utilities::PathHelper
+
+  let(:project) { FactoryGirl.create(:project_with_types, id: 5, is_public: false) }
+
+  describe '#post' do
+    let(:status) { FactoryGirl.build(:status, is_default: true) }
+    let(:priority) { FactoryGirl.build(:priority, is_default: true) }
+    let(:parameters) { { subject: 'new work packages' } }
+    let(:permissions) { [:add_work_packages, :view_project] }
+    let(:role) { FactoryGirl.create(:role, permissions: permissions) }
+    let(:current_user) do
+      FactoryGirl.build(:user, member_in_project: project, member_through_role: role)
+    end
+    let(:path) { api_v3_paths.work_packages_by_project project.id }
+
+    before do
+      status.save!
+      priority.save!
+
+      allow(User).to receive(:current).and_return current_user
+      ActionMailer::Base.deliveries.clear
+      post path, parameters.to_json, 'CONTENT_TYPE' => 'application/json'
+    end
+
+    context 'notifications' do
+      let(:permissions) { [:add_work_packages, :view_project, :view_work_packages] }
+
+      it 'sends a mail by default' do
+        post path, parameters.to_json, 'CONTENT_TYPE' => 'application/json'
+        expect(ActionMailer::Base.deliveries.count).to eq(2)
+      end
+
+      context 'without notifications' do
+        let(:path) { "#{api_v3_paths.work_packages_by_project(project.id)}?notify=false" }
+
+        it 'should not send a mail' do
+          post path, parameters.to_json, 'CONTENT_TYPE' => 'application/json'
+          expect(ActionMailer::Base.deliveries.count).to eq(0)
+        end
+      end
+    end
+
+    it 'should return Created(201)' do
+      expect(last_response.status).to eq(201)
+    end
+
+    it 'should create a work package' do
+      expect(WorkPackage.all.count).to eq(1)
+    end
+
+    it 'should use the given parameters' do
+      expect(WorkPackage.first.subject).to eq(parameters[:subject])
+    end
+
+    context 'unauthorized user' do
+      let(:current_user) { FactoryGirl.build(:user) }
+
+      it 'should hide the endpoint' do
+        expect(last_response.status).to eq(404)
+      end
+    end
```

Comment:
This is disturbingly irritating.

The reason for the 404 is that the project is not visible,

however there is a difference between the missing permission to view the project (404) and to add work packages (403), the latter case not being covered.

You should differentiate both cases, it should suffice to change the `let(:permissions)` per context.